File: hdf-converters/shared/go/bom/spdx.go
// SPDX SBOM -> normalized HDF BillOfMaterials.
//
// Maps packages[] to normalized SBOM packages. purl comes from an externalRefs
// entry with referenceType "purl"; licenses come from licenseConcluded /
// licenseDeclared with NOASSERTION/NONE sentinels filtered out.

package bom

import shared "github.com/mitre/hdf-libs/hdf-converters/v3/shared/go"

var spdxLicenseFields = []string{"licenseConcluded", "licenseDeclared"}

// purlFromExternalRefs returns the first externalRefs[].referenceLocator whose
// referenceType is "purl".
func purlFromExternalRefs(refs any) string {
	entries, ok := refs.([]any)
	if !ok {
		return ""
	}
	for _, ref := range entries {
		r := asRecord(ref)
		if r != nil && r["referenceType"] == "purl" {
			if locator := asString(r["referenceLocator"]); locator != "" {
				return locator
			}
		}
	}
	return ""
}

func extractSPDXLicenses(pkg map[string]any) []string {
	out := []string{}
	for _, field := range spdxLicenseFields {
		license := cleanLicense(pkg[field])
		if license == "" {
			continue
		}
		if !contains(out, license) {
			out = append(out, license)
		}
	}
	return out
}

func contains(list []string, value string) bool {
	for _, item := range list {
		if item == value {
			return true
		}
	}
	return false
}

func packageToPackage(source any) (SBOMPackage, bool) {
	p := asRecord(source)
	name := asString(p["name"])
	if p == nil || name == "" {
		return SBOMPackage{}, false
	}

	pkg := SBOMPackage{Name: name}
	if version := asString(p["versionInfo"]); version != "" {
		pkg.Version = strPtr(version)
	}
	if purl := purlFromExternalRefs(p["externalRefs"]); purl != "" {
		pkg.Purl = strPtr(purl)
	}
	if licenses := extractSPDXLicenses(p); len(licenses) > 0 {
		pkg.Licenses = licenses
	}

	enrichFromPurl(&pkg)
	return pkg, true
}

// ParseSPDX normalizes an SPDX SBOM object into an sbom BillOfMaterials.
func ParseSPDX(obj map[string]any) *NormalizedBom {
	sourcePackages, _ := obj["packages"].([]any)
	packages := []SBOMPackage{}
	for _, source := range sourcePackages {
		if pkg, ok := packageToPackage(source); ok {
			packages = append(packages, pkg)
		}
	}

	return normalized(BuildBom(BuildBomParts{
		BOMType:  BOMTypeSbom,
		Format:   FormatSPDX,
		Packages: shared.LimitSliceWithWarning(packages, maxPackages, "package"),
		UniqueID: strPtr(asString(obj["documentNamespace"])),
	}))
}

File: hdf-converters/shared/typescript/bom/spdx.ts
/**
 * SPDX SBOM -> normalized HDF BillOfMaterials.
 *
 * Maps packages[] to normalized SBOM packages. purl comes from an externalRefs
 * entry with referenceType 'purl'; licenses come from licenseConcluded /
 * licenseDeclared with NOASSERTION/NONE sentinels filtered out.
 */

import {
  BOMType,
  type NormalizedBom,
  type SBOMPackage,
} from './model.js';
import {
  asRecord,
  asString,
  buildBom,
  cleanLicense,
  enrichFromPurl,
} from './normalize.js';
import { limitArrayWithWarning } from '../converterutil.js';

const SPDX_LICENSE_FIELDS = ['licenseConcluded', 'licenseDeclared'] as const;

/** First externalRefs[].referenceLocator whose referenceType is 'purl'. */
function purlFromExternalRefs(refs: unknown): string | undefined {
  if (!Array.isArray(refs)) return undefined;
  for (const ref of refs) {
    const r = asRecord(ref);
    if (r?.referenceType === 'purl') {
      const locator = asString(r.referenceLocator);
      if (locator) return locator;
    }
  }
  return undefined;
}

function extractLicenses(pkg: Record<string, unknown>): string[] {
  const out: string[] = [];
  for (const field of SPDX_LICENSE_FIELDS) {
    const license = cleanLicense(pkg[field]);
    if (license && !out.includes(license)) out.push(license);
  }
  return out;
}

function packageToPackage(source: unknown): SBOMPackage | undefined {
  const p = asRecord(source);
  const name = p ? asString(p.name) : undefined;
  if (!p || !name) return undefined;

  const pkg: SBOMPackage = { name };
  const version = asString(p.versionInfo);
  if (version) pkg.version = version;
  const purl = purlFromExternalRefs(p.externalRefs);
  if (purl) pkg.purl = purl;
  const licenses = extractLicenses(p);
  if (licenses.length > 0) pkg.licenses = licenses;

  enrichFromPurl(pkg);
  return pkg;
}

export function parseSPDX(obj: Record<string, unknown>): NormalizedBom {
  const sourcePackages = Array.isArray(obj.packages) ? obj.packages : [];
  const packages: SBOMPackage[] = [];
  for (const source of sourcePackages) {
    const pkg = packageToPackage(source);
    if (pkg) packages.push(pkg);
  }

  const uniqueId = asString(obj.documentNamespace);
  return buildBom({
    bomType: BOMType.Sbom,
    format: 'spdx',
    packages: limitArrayWithWarning(packages, 'package'),
    uniqueId,
  });
}
